### PR TITLE
Compare it equals

### PR DESCRIPTION
**Pull request by thiagotalma**
@cebe , how about stop this old thread #2790?

A parameter solves the problem.

### CODE REVIEW

**Comment by thiagotalma** [comment]
@qiangxue , Please give your opinion here.

File: framework/db/BaseActiveRecord.php [hunk 1]
```diff
@@ -514,12 +514,13 @@ public function markAttributeDirty($name)
...
-     * @return boolean whether the attribute has been changed
+     * @param bool $identical if the comparison is made as identical
+     * @return bool whether the attribute has been changed
      */
-    public function isAttributeChanged($name)
+    public function isAttributeChanged($name, $identical = true)
     {
         if (isset($this->_attributes[$name], $this->_oldAttributes[$name])) {
-            return $this->_attributes[$name] !== $this->_oldAttributes[$name];
+            return $identical ? $this->_attributes[$name] !== $this->_oldAttributes[$name] : $this->_attributes[$name] != $this->_oldAttributes[$name];
```
Comment (cebe) [inline framework/db/BaseActiveRecord.php]: code looks confusing on one line, should be a normal if `if ($identical) ... else ...`

**Comment by qiangxue** [comment]
What's the use of this parameter? It won't fix #2790.

File: framework/db/BaseActiveRecord.php (hunk 1 above)
Comment (thiagotalma) [inline framework/db/BaseActiveRecord.php]: Ok @cebe

**Comment by thiagotalma** [comment]
@qiangxue the intention is not to solve the issue, but provide a way to bypass it when necessary.

**Comment by qiangxue** [comment]
> ...but provide a way to bypass it when necessary.

In which use case?

**Comment by thiagotalma** [comment]
```php
public function beforeSave($insert)
{
    if (!$insert && $this->isAttributeChanged('user_id', false)) {
        // Do something
    }

    return parent::beforeSave($insert);
}
```

@qiangxue 
Consider that user_id is a field indicating a relationship. You need to perform a procedure if the relationship changes.

Without modification of PR, even if you do not change a field it will always be considered changed because the comparison as identical.

**Comment by qiangxue** [comment]
Convinced. Could you add a line to the CHANGELOG file? Thanks.

**Comment by qiangxue** [comment]
thanks!

**Comment by samdark** [comment]
But isn't it breaking backwards compatibility?

**Comment by klimov-paul** [comment]
No, by default `!==` used just like it was before.

**Comment by samdark** [comment]
I mean signature break.

**Comment by klimov-paul** [comment]
Yes, if someone is overriding this method he will receive E_STRICT.

**Comment by samdark** [comment]
Should be mentioned in upgrade guide then.

**Comment by cebe** [comment]
I'm afraid we are missing this kind of break in some other places. e.g. here: https://github.com/yiisoft/yii2-framework/compare/2.0.3...master#diff-966df45f15a870463a3012e393f2c25dR247 have not scrolled further yet.

**Pull request merged**